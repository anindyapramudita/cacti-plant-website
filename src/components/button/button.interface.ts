import { ButtonHTMLAttributes } from "react";


export interface IButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>{
    children: string;
    color?: 'primary' | 'secondary';
    variant?: 'filled' | 'reversed' | 'outlined';
    size?: 'small' | 'medium' | 'large';
    isLoading?: boolean;
    disabled?: boolean;
    fullWidth?: boolean;
    loaderDot?: number;
    loaderSize?: 'mini' | 'small' | 'medium' | 'large';
  }