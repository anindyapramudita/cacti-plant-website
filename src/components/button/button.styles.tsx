import { ralewayThin } from "@/styles/fonts";
import styled from "styled-components";

export const StylesWrapper = styled.button`
  ${ralewayThin}

  --button-padding: 8px 24px;
  --button-font-size: 1rem;
  --button-background-color: var(--new-primary);
  --button-color: var(--white);
  --button-border-color: var(--new-primary);
  --button-border: 0.0625rem solid var(--button-border-color);

  cursor: pointer;
  display: flex;
  gap: 5px;
  align-items: center;
  justify-content: center;
  line-height: 1;
  transition: background-color 0.4s ease, border-color 0.4s ease;
  background-color: var(--button-background-color);
  border: var(--button-border);
  padding: var(--button-padding);
  color: var(--button-color);
  font-size: var(--button-font-size);
  border-radius: 8px;

  &:hover {
    color: var(--button-color);
    --button-background-color: var(--new-primary-dark);
    --button-border: 0.0625rem solid var(--new-primary-dark);
  }

  &.button-filled-secondary {
    --button-background-color: var(--new-secondary);
    --button-border: 0.0625rem solid var(--new-secondary);
    &:hover {
      --button-background-color: var(--new-secondary-dark);
      --button-border: 0.0625rem solid var(--new-secondary-dark);
    }
  }

  &.button-filled-monochrome {
    --button-background-color: var(--black);
    --button-border: 0.0625rem solid var(--black);
    &:hover {
      --button-background-color: var(--very-dark-grey);
      --button-border: 0.0625rem solid var(--very-dark-grey);
    }
  }

  &.button-outlined-primary,
  &.button-reversed-primary {
    --button-background-color: var(--white);
    --button-color: var(--new-primary);

    &:hover {
      --button-background-color: var(--new-primary);
      --button-border: 0.0625rem solid var(--new-primary);
      --button-color: var(--white);
    }
  }

  &.button-outlined-secondary,
  &.button-reversed-secondary {
    --button-background-color: var(--white);
    --button-color: var(--new-secondary);

    &:hover {
      --button-background-color: var(--new-secondary);
      --button-border: 0.0625rem solid var(--new-secondary);
      --button-color: var(--white);
    }
  }

  &.button-outlined-monochrome,
  &.button-reversed-monochrome {
    --button-background-color: var(--white);
    --button-border: 0.0625rem solid var(--very-dark-grey);
    --button-color: var(--black);

    &:hover {
      --button-background-color: var(--very-dark-grey);
      --button-border: 0.0625rem solid var(--very-dark-grey);
      --button-color: var(--white);
    }
  }

  &.button-outlined-secondary {
    --button-border: 0.0625rem solid var(--new-secondary);

    &:hover {
      --button-color: var(--white);
    }
  }

  &.button-reversed-primary {
    --button-border: 0.0625rem solid var(--white);
  }

  &.button-reversed-secondary {
    --button-border: 0.0625rem solid var(--white);
  }

  &.size-small {
    --button-padding: 0.625rem 16px;
    --button-font-size: 0.875rem;
  }

  &.size-large {
    --button-padding: 0.7125em 3.0125em;
    --button-font-size: 1.125rem;
  }

  &.button-fullwidth {
    width: 100%;
  }

  &:disabled {
    --button-background-color: var(--disabled-light);
    --button-border: 0.0625rem solid var(--disabled-light);
    --button-color: var(--disabled-dark);
    cursor: default;
    position: relative;

    &:hover {
      --button-background-color: var(--disabled-light);
      --button-border: 0.0625rem solid var(--disabled-light);
      --button-color: var(--disabled-dark);
      cursor: default;
    }
  }

  &.button-loading {
    position: relative;
    --button-color: var(--black);

    color: transparent;

    .button-icon {
      visibility: hidden;
    }

    &:hover {
      --button-background-color: var(--disabled-light);
      --button-border: 0.0625rem solid var(--disabled-light);
      cursor: default;
    }
  }
`;
