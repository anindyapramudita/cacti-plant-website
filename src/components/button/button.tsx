import { FC } from "react";
import { Loader } from "../loader";
import { IButtonProps } from "./button.interface";
import { StylesWrapper, StylesText } from "./button.styles";

export const Button: FC<IButtonProps> = ({
  children,
  disabled = false,
  isLoading = false,
  variant = "filled",
  color = "primary",
  size = "medium",
  fullWidth = false,
  loaderDot = 3,
  loaderSize = "small",
  ...props
}) => {
  if (!children) {
    return null;
  }

  return (
    <StylesWrapper
      data-testid="button"
      color={color}
      variant={variant}
      disabled={disabled || isLoading}
      isLoading={isLoading}
      size={size}
      fullWidth={fullWidth}
      {...props}
    >
      <p>{children}</p>
      {isLoading && <Loader size={loaderSize} number={loaderDot} />}
    </StylesWrapper>
  );
};
