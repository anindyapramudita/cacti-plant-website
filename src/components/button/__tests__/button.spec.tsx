/* eslint-disable no-undef */
import { render, fireEvent, act } from "@testing-library/react";
import { Button } from "../button";
import { BsGoogle } from "react-icons/bs";

const text = "Test Button";

describe("Button Component", () => {
  test("Button should render with the right properties", () => {
    const { getByTestId, getByText } = render(<Button>{text}</Button>);

    const button = getByTestId("button");

    const buttonLabel = getByText(text);
    const loader = button.querySelector('[data-testid="loader"]');

    expect(button).toBeInTheDocument();
    expect(buttonLabel).toBeInTheDocument();

    expect(button).toHaveClass("button-filled-primary size-medium");
    expect(loader).not.toBeInTheDocument();
    expect(getComputedStyle(button)).toMatchSnapshot();
  });

  test("Button should render the right variant and type", () => {
    const { getByTestId } = render(
      <Button variant="outlined" color="secondary" size="small">
        {text}
      </Button>
    );

    const button = getByTestId("button");

    expect(button).toBeInTheDocument();

    expect(button).toHaveClass("button-outlined-secondary size-small");
  });

  test("Button should render icon if props has icon", () => {
    const { getByTestId } = render(
      <Button icon={<BsGoogle data-testid="icon-button" />}>{text}</Button>
    );

    const icon = getByTestId("icon-button");

    expect(icon).toBeInTheDocument();
  });

  test("Button should call onClick handler when triggered", () => {
    const onClickFn = jest.fn();
    const { getByTestId } = render(<Button onClick={onClickFn}>{text}</Button>);

    const button = getByTestId("button");

    act(() => {
      fireEvent.click(button);
    });

    expect(onClickFn).toBeCalledTimes(1);
  });

  test("Button should display loader when button is loading", () => {
    const { getByTestId } = render(<Button isLoading={true}>{text}</Button>);

    const button = getByTestId("button");
    const loader = button.querySelector('[data-testid="loader"]');

    expect(button).toBeDisabled();
    expect(loader).toBeInTheDocument();
  });

  test("Button should be disabled when loading is true", () => {
    const onClickFn = jest.fn();

    const { getByTestId } = render(
      <Button isLoading={true} onClick={onClickFn}>
        {text}
      </Button>
    );

    const button = getByTestId("button");

    act(() => {
      fireEvent.click(button);
    });

    expect(button).toBeDisabled();
    expect(onClickFn).not.toBeCalled();
  });

  test("Button should be disabled when disabled prop is true", () => {
    const onClickFn = jest.fn();

    const { getByTestId } = render(
      <Button disabled onClick={onClickFn}>
        {text}
      </Button>
    );

    const button = getByTestId("button");

    act(() => {
      fireEvent.click(button);
    });

    expect(button).toBeDisabled();
    expect(onClickFn).not.toBeCalled();
  });

  test("Button shouldn't render if children is undefined", () => {
    const { container } = render(<Button>{undefined}</Button>);

    const button = container.querySelector('[data-testid="button"]');
    expect(button).not.toBeInTheDocument();
  });
});
