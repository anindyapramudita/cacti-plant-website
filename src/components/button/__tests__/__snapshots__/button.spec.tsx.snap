// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Button should render with the right properties 1`] = `
CSSStyleDeclaration {
  "0": "font",
  "1": "text-rendering",
  "10": "align-items",
  "11": "cursor",
  "12": "color",
  "13": "padding",
  "14": "border",
  "15": "background-color",
  "16": "box-sizing",
  "17": "--font-size",
  "18": "font-weight",
  "19": "--button-padding",
  "2": "letter-spacing",
  "20": "--button-font-size",
  "21": "--button-background-color",
  "22": "--button-color",
  "23": "--button-border-color",
  "24": "--button-border",
  "25": "transition",
  "26": "visibility",
  "3": "word-spacing",
  "4": "line-height",
  "5": "text-transform",
  "6": "text-indent",
  "7": "text-shadow",
  "8": "display",
  "9": "text-align",
  "_importants": {
    "--button-background-color": "",
    "--button-border": "",
    "--button-border-color": "",
    "--button-color": "",
    "--button-font-size": "",
    "--button-padding": "",
    "--font-size": "",
    "align-items": "",
    "background-color": "",
    "border": "",
    "box-sizing": "",
    "color": "",
    "cursor": "",
    "display": "",
    "font": "",
    "font-size": "",
    "font-weight": "",
    "letter-spacing": "",
    "line-height": "",
    "margin": "",
    "padding": "",
    "text-align": "",
    "text-indent": "",
    "text-rendering": "",
    "text-shadow": "",
    "text-transform": "",
    "transition": "",
    "visibility": undefined,
    "word-spacing": "",
  },
  "_length": 27,
  "_onChange": [Function],
  "_values": {
    "--button-background-color": "var(--primary)",
    "--button-border": "0.0625rem solid var(--button-border-color)",
    "--button-border-color": "var(--primary)",
    "--button-color": "var(--white)",
    "--button-font-size": "1rem",
    "--button-padding": "0.6875em 3em",
    "--font-size": "1rem",
    "align-items": "flex-start",
    "background-color": "ButtonFace",
    "border": "2px outset buttonface",
    "border-bottom": "2px outset buttonface",
    "border-bottom-color": "buttonface",
    "border-bottom-style": "outset",
    "border-bottom-width": "2px",
    "border-color": "buttonface",
    "border-left": "2px outset buttonface",
    "border-left-color": "buttonface",
    "border-left-style": "outset",
    "border-left-width": "2px",
    "border-right": "2px outset buttonface",
    "border-right-color": "buttonface",
    "border-right-style": "outset",
    "border-right-width": "2px",
    "border-style": "outset",
    "border-top": "2px outset buttonface",
    "border-top-color": "buttonface",
    "border-top-style": "outset",
    "border-top-width": "2px",
    "border-width": "2px",
    "box-sizing": "border-box",
    "color": "ButtonText",
    "cursor": "pointer",
    "display": ""inline-block"",
    "font": "-webkit-small-control",
    "font-family": "-webkit-small-control",
    "font-weight": "300",
    "letter-spacing": "normal",
    "line-height": "1",
    "padding": "2px 6px 3px 6px",
    "padding-bottom": "3px",
    "padding-left": "6px",
    "padding-right": "6px",
    "padding-top": "2px",
    "text-align": "center",
    "text-indent": "0",
    "text-rendering": "auto",
    "text-shadow": "none",
    "text-transform": "none",
    "transition": "background-color 0.4s ease,border-color 0.4s ease",
    "visibility": "visible",
    "word-spacing": "normal",
  },
}
`;
